gh-118209: Add structured exception handling to mmap module

Dobatymo opened the pull request:
If `DONT_USE_SEH` is not defined, catch page errors and access violations when reading/writing the mmap pointer and convert it into a Python exception.

Methods done:
- `mmap_read_method`
- `mmap_read_byte_method`
- `mmap_read_line_method`
- `mmap_write_method`
- `mmap_write_byte_method`
- `mmap_item`
- `mmap_subscript`
- `mmap_ass_item`
- `mmap_ass_subscript`
- `mmap_move_method`
- `mmap_gfind` (hopefully I didn't miss any resource handling here)

Other comments:
- `LsaNtStatusToWinError` is equivalent to `RtlNtStatusToDosError` (see comment below). So I am using this since it's easier to call.
- one other tiny bug is fixed in `mmap_read_line_method`. Previously the file pointer would still be advanced in case of a memory allocation error.

Issues with the PR:
- other methods should be handled as well, but I want to wait on some feedback first
  - `mmap_resize_method`

I don't know how to fix:
- `mmap_buffer_getbuf`: this creates a buffer object without actually reading from memory (yet). This will cause issues when the buffer is read in other parts of the code. The only way to fix this would be make all buffer methods SEH aware also... :(

See
- https://learn.microsoft.com/en-us/windows/win32/api/memoryapi/nf-memoryapi-mapviewoffile#remarks
- https://learn.microsoft.com/en-us/windows/win32/memory/reading-and-writing-from-a-file-view
- https://learn.microsoft.com/en-us/windows/win32/api/winnt/ns-winnt-exception_record

<!-- gh-issue-number: gh-118209 -->
* Issue: gh-118209
<!-- /gh-issue-number -->

Dobatymo committed:
add structured exception handling to mmap_read_method

blurb-it[bot] committed:
📜🤖 Added by blurb_it.

Dobatymo committed:
adjust indentation

zooba commented on Modules/mmapmodule.c:
```diff
@@ -291,6 +291,17 @@ mmap_read_line_method(mmap_object *self,
     return result;
 }
 
+#if defined(MS_WIN32) && !defined(DONT_USE_SEH)
```
We should use `MS_WINDOWS` rather than `MS_WIN32`.

Where does `DONT_USE_SEH` come from? I don't recall this being a standard preprocessor definition. If it's our own, it should either be a (positive) `HAVE_SEH` that is defined in pyconfig.h, or a `Py_DONT_USE_SEH` to indicate that it's Python specific.

Dobatymo replied:
I was looking at existing code which uses SEH. Like https://github.com/python/cpython/blob/8227883d1f1bbb6560e5f175d7ee49f013c094bd/Modules/_ctypes/callproc.c#L827-L984

zooba replied:
I forget the exact reason, but ctypes only uses SEH on 32-bit Windows. Possibly because the x64/ARM64 calling conventions don't allow recovery across separately compiled modules? Or there's a reliance on compiler-specific metadata that isn't always there? Whatever it is, that shouldn't matter for this scenario.

While MSVC is our only supported compiler that will set `MS_WINDOWS`, people sometimes use Clang to compile for Windows (and I guess GCC, but I haven't heard of anyone doing that recently). I presume those don't support `__try`, but it's probably best to omit it based on compiler detection rather than a separate flag.

(Also, `ctypes` is... ehm... not the best example of good coding conventions 😉 )

Dobatymo replied:
So I just check for MS_WINDOWS and ignore MS_WIN32 and DONT_USE_SEH?

eryksun replied:
`MS_WIN32` is always defined on Windows. Restricting to 64-bit requires checking for `MS_WIN64`. ctypes uses SEH on all supported versions of Windows as long as `DONT_USE_SEH` is not defined.

zooba commented on Modules/mmapmodule.c:
```diff
@@ -307,8 +318,23 @@ mmap_read_method(mmap_object *self,
+#if defined(MS_WIN32) && !defined(DONT_USE_SEH)
+    EXCEPTION_RECORD record;
+    __try {
+        result = PyBytes_FromStringAndSize(&self->data[self->pos], num_bytes);
```
I'd feel more confident about recovering from this error if we `PyBytes_FromStringAndSize(NULL, num_bytes)` outside the try/except, then do the copy inside. That way we can correctly free any memory that was allocated.

Dobatymo replied:
You are right. The current implementation will not free allocated memory correctly.

Dobatymo replied:
If we do `PyBytes_FromStringAndSize(NULL, num_bytes)` first and then `memcpy` inside the try, we will lose the one-character optimization provided by `PyBytes_FromStringAndSize`. If we want to keep it, we would have to copy and paste more parts of the function body. Or create a separate `_PyBytes_FromStringAndSizeSehSafe` function.

https://github.com/python/cpython/blob/8227883d1f1bbb6560e5f175d7ee49f013c094bd/Objects/bytesobject.c#L120-L123

zooba replied:
Perhaps then we want a `safe_read_byte(ptr, &out_char)` function to do the SEH and then special-case the entire call for `num_bytes == 1`? That's probably better overall than trying to handle things inside `safe_memcpy` anyway.

zooba reviewed (commented):
PR looks good so far, though as this is a draft I'll leave it with you for now.

One other possibility may be to protect `PyBytes_FromString` internally, which could cut off an entire class of crashes instead of just one. Thoughts?

eryksun commented on Modules/mmapmodule.c:
```diff
@@ -307,8 +318,23 @@ mmap_read_method(mmap_object *self,
+        self->pos += num_bytes;
+    }
+    __except (HandlePageException(GetExceptionInformation(), &record)) {
+        NTSTATUS code = record.ExceptionInformation[2];
```
`OSError` doesn't support `NTSTATUS` codes. You'll have to dynamically call `ULONG RtlNtStatusToDosError(NTSTATUS Status)` from the system library "ntdll.dll".

zooba replied:
Or we can raise a generic `OSError` and format the error code into the string. It's incredibly unlikely that the caller can actually handle any specific errors, so all that is likely to happen is logging and cleanup. We don't need to get winerror perfect for those.

Dobatymo committed:
restructure to use safe_memcpy to avoid memory leak

eryksun commented on Modules/mmapmodule.c:
```diff
@@ -255,6 +255,44 @@ do {                                                                    \
 
+#if defined(MS_WIN32) && !defined(DONT_USE_SEH)
+static DWORD
+HandlePageException(EXCEPTION_POINTERS *ptrs, EXCEPTION_RECORD *record)
```
```suggestion
filter_page_exception(EXCEPTION_POINTERS *ptrs, EXCEPTION_RECORD *record)
```

eryksun commented on Modules/mmapmodule.c:
```diff
@@ -255,6 +255,44 @@ do {                                                                    \
+        memcpy(dest, src, count);
+        return 0;
+    }
+    __except (HandlePageException(GetExceptionInformation(), &record)) {
```
```suggestion
    __except (filter_page_exception(GetExceptionInformation(), &record)) {
```

Dobatymo committed:
Apply suggestions from code review

rename HandlePageException to filter_page_exception

Co-authored-by: Eryk Sun <eryksun@gmail.com>

zooba commented on Modules/mmapmodule.c:
```diff
@@ -307,8 +352,17 @@ mmap_read_method(mmap_object *self,
+    if (result == NULL) {
+        return NULL;
+    }
+    if (safe_memcpy(((PyBytesObject *) result)->ob_sval, self->data + self->pos, num_bytes) < 0) {
```
```suggestion
    if (safe_memcpy(PyBytes_AS_STRING(result), self->data + self->pos, num_bytes) < 0) {
```

zooba commented on Modules/mmapmodule.c:
```diff
@@ -264,7 +302,14 @@ mmap_read_byte_method(mmap_object *self,
     }
-    return PyLong_FromLong((unsigned char)self->data[self->pos++]);
+    unsigned char dest;
+    if (safe_memcpy(dest, self->data + self->pos, 1) < 0) {
```
```suggestion
    if (safe_memcpy(&dest, self->data + self->pos, 1) < 0) {
```

zooba reviewed (commented):
I like how this is looking. Let us know when you'd like CI run and we can hit the button to get it going, and feel free to start adapting other cases in this file.

Dobatymo committed:
fix passing dest as pointer in mmap_read_byte_method

Co-authored-by: Steve Dower <steve.dower@microsoft.com>

Dobatymo committed:
use PyBytes_AS_STRING instead of raw access

Co-authored-by: Steve Dower <steve.dower@microsoft.com>

Dobatymo committed:
handle mmap_write_method

Dobatymo committed:
handle mmap_write_byte_method

Dobatymo committed:
add safe_byte_copy, safe_memmove, safe_memchr, safe_copy_from_slice and safe_copy_to_slice

handle SEH in mmap_read_line_method, mmap_item, mmap_ass_item and mmap_subscript, mmap_ass_subscript, mmap_move_method

zooba commented on Modules/mmapmodule.c:
```diff
@@ -293,6 +303,69 @@ safe_memcpy(void *restrict dest, const void *restrict src, size_t count) {
 }
 
+#if defined(MS_WIN32) && !defined(DONT_USE_SEH)
+#define handle_invalid_mem(sourcecode) \
```
This is clever! I haven't seen it done like this before. A few suggestions:
* can we make it `HANDLE_INVALID_MEM` so it's more obviously a macro
* can we use `{` `}` around the code when it's used to make it more obviously a block
* can we put the `return 0` in the functions rather than the macro (return -1 here is fine)

zooba commented on Modules/mmapmodule.c:
```diff
@@ -312,36 +385,67 @@ mmap_read_byte_method(mmap_object *self,
+            return NULL;
+        }
+        else {
+            PyBytesObject *op = CHARACTER(dest & 255);
```
Is having this here really an improvement over `PyBytes_FromStringAndSize(&dest, 1)`? Performance wise, probably, but I don't like having performance/memory optimizations like this being replicated throughout the codebase.

Dobatymo replied:
Since this optimization is present in `PyBytes_FromStringAndSize` https://github.com/python/cpython/blob/8227883d1f1bbb6560e5f175d7ee49f013c094bd/Objects/bytesobject.c#L120-L123 but is never used when passing NULL I added it here since you mentioned above that single byte performance is important in regard to memcpy.

zooba replied:
Agreed, I'm proposing passing `&dest` instead of `NULL` so we'll still get the optimisation but without duplicating the code.

Maintainability is also important, as well as performance. And the performance will always be better for callers who read more than one byte at a time, so unless we're talking a 50-100% difference on a microbenchmark, I think maintainability wins (the `NULL` behaviour includes dynamic memory allocation, so is _definitely_ going to be much slower than passing `&dest`).

Dobatymo replied:
@zooba Please have a look at my latest commit if that's the way you meant to implement it.

zooba reviewed (commented):
Love that macro trick! I had never considered passing lines of code as an argument before.

zooba commented on Modules/mmapmodule.c:
```diff
@@ -264,7 +375,44 @@ mmap_read_byte_method(mmap_object *self,
+    }
+    else {
+        self->pos++;
+        return PyLong_FromLong(dest);
```
```suggestion
        return PyLong_FromLong((unsigned char)dest);
```

For the failing test.

zooba commented on Modules/mmapmodule.c:
```diff
@@ -273,30 +421,31 @@ mmap_read_line_method(mmap_object *self,
     start = self->data + self->pos;
-    eol = memchr(start, '\n', remaining);
+
+    if (safe_memchr(&eol, start, '\n', remaining) < 0) {
```
```suggestion
    if (safe_memchr((void **)&eol, start, '\n', remaining) < 0) {
```

For the compiler warnings (I hope... I didn't test whether this will actually fix them)

Dobatymo replied:
I changed the signature of `safe_memchr` to use `char **` instead. Otherwise I am dereferencing a void pointer in the function.

zooba commented on Modules/mmapmodule.c:
```diff
@@ -1097,8 +1268,13 @@ mmap_ass_subscript(mmap_object *self, PyObject *item, PyObject *value)
-        self->data[i] = (char) v;
-        return 0;
+
+        if (safe_byte_copy(self->data + i, (char *) &v) < 0) {
```
```suggestion
        char v_char = (char) v;
        if (safe_byte_copy(self->data + i, &v_char) < 0) {
```

There's a subtle difference between the casts, so safest to keep casting the value to `char` and then reference the address.

Dobatymo committed:
improve macro

fix some type issues

Dobatymo committed:
Update Modules/mmapmodule.c

cast to char and then deref instead of casting (char *)

Co-authored-by: Steve Dower <steve.dower@microsoft.com>

zooba commented on Modules/mmapmodule.c:
```diff
@@ -255,6 +258,120 @@ do {                                                                    \
+#endif
+
+int
+safe_memcpy(void *restrict dest, const void *restrict src, size_t count) {
```
This one can be defined using the macro as well, right?

I'm reading this section from the POV of someone who develops on Linux, trying to figure out where to add the code to make Linux work as well. If that can be made obvious (perhaps with a `// TODO: ...` comment), it's more likely we'll get someone come and add it.

Dobatymo replied:
Yes it can be defined using the macro too. So far I kept it like this because I had this optional check for zero length inputs. This check is either totally unnecessary though or doesn't hurt to be part of the non-windows implementation.

Dobatymo committed:
handle EXCEPTION_ACCESS_VIOLATION

define safe_memcpy in terms of HANDLE_INVALID_MEM

Dobatymo committed:
fix formatting

Dobatymo committed:
use PyBytes_FromStringAndSize instead of extracting the optization

Dobatymo committed:
update news entry

zooba commented on Misc/NEWS.d/next/Windows/2024-04-24-05-16-32.gh-issue-118209.Ryyzlz.rst:
```diff
@@ -0,0 +1,12 @@
+* :func:`mmap.__getitem__`
+* :func:`mmap.__setitem__`
+
+on Windows when the underlying file operation raises errors or an access violation occurs.
```
```suggestion
Avoid crashing in :mod:`mmap` on Windows when the mapped memory is inaccessible
due to file system errors or access violations.
```

Let's just keep this simple, as it's going to be in a very long list of changes.

However, in `Doc/whatsnew/3.13.rst` you'll find a section for "Improved modules" that contains "mmap" (or can be added), and there you can add something like this:

```
* :class:`mmap.mmap` is now protected from crashing on Windows when the mapped memory
  is inaccessible due to file system errors or access violations.
  (Contributed by <your name here> in :gh:`118209`)
```

zooba replied:
It's probably worth adding a note to `Doc/library/mmap.rst` saying that "most operations on Windows are protected from crashing due to file system errors or access violations, but may raise ``OSError`` if one of these occurs".

I don't think we need to necessarily point at each individual function.

Dobatymo committed:
Update Misc/NEWS.d/next/Windows/2024-04-24-05-16-32.gh-issue-118209.Ryyzlz.rst

Co-authored-by: Steve Dower <steve.dower@microsoft.com>

Dobatymo committed:
updated whatsnew entry

Dobatymo committed:
fix typo

Dobatymo committed:
fix handling some cases in mmap_subscript

Dobatymo committed:
add test using private function

eryksun commented on Lib/test/test_mmap.py:
```diff
@@ -1058,6 +1058,44 @@ def __exit__(self, exc_type, exc_value, traceback):
+    def test_access_violations(self):
+        PAGE_NOACCESS = 0x01
+
+        with open(TESTFN, 'bw+') as f:
```
A child process is needed in these tests because a regression could cause the process to crash. You should be able to use `assert_python_ok('-c', code)`, maybe with individual tests. Import it from `test.support.script_helper`.

eryksun commented on Lib/test/test_mmap.py:
```diff
@@ -1058,6 +1058,44 @@ def __exit__(self, exc_type, exc_value, traceback):
                 with self.assertRaisesRegex(ValueError, "mmap closed or invalid"):
                     m.write_byte(X())
 
+    @unittest.skipUnless(os.name == 'nt', 'requires Windows')
```
```suggestion
    @unittest.skipUnless(os.name == 'nt', 'requires Windows')
    @unittest.skipUnless(hasattr(mmap.mmap, '_protect'), 'test needs debug build')
```

Dobatymo committed:
Update Lib/test/test_mmap.py

Co-authored-by: Eryk Sun <eryksun@gmail.com>

Dobatymo committed:
run test in subprocess

eryksun commented on Lib/test/test_mmap.py:
```diff
@@ -1058,6 +1060,59 @@ def __exit__(self, exc_type, exc_value, traceback):
+                    list(m)  # test mmap_item
+        """)
+        # stderr includes "Windows fatal exception: access violation" logs,
+        # even when the access violations are correctly transformed into exceptions
```
Call `faulthandler.disable()` at the start of the executed code. The issue is that mmap uses a stack handler, and vectored handlers, such as the one set by faulthandler, are called first.

Dobatymo committed:
add expected failure for find method

Dobatymo committed:
disable faulthandler

Dobatymo committed:
typo

Dobatymo committed:
trailing blanks

eryksun commented on Lib/test/test_mmap.py:
```diff
@@ -1058,6 +1060,90 @@ def __exit__(self, exc_type, exc_value, traceback):
+            import os
+            from contextlib import suppress
+
+            faulthandler.disable()
```
```suggestion
            # Prevent logging access violations to stderr.
            faulthandler.disable()
```
It's helpful to explain why this step is required.

Dobatymo committed:
Update Lib/test/test_mmap.py

Co-authored-by: Eryk Sun <eryksun@gmail.com>

eryksun commented on Lib/test/test_mmap.py:
```diff
@@ -1058,6 +1060,91 @@ def __exit__(self, exc_type, exc_value, traceback):
+                with suppress(OSError):
+                    m.find(b'A')
+        """)
+        assert_python_ok("-c", code)
```
I don't see why a test is needed for the `find()` method if it hasn't been made safe. Are you planning to support `[r]find()`? Anyway, this test could use `assert_python_failure()`. Then it wouldn't need `unittest.expectedFailure`.

Dobatymo replied:
I added the test basically as an anti-test to have higher confidence that the other test is working as it's supposed to.

Dobatymo replied:
and for `expectedFailure`. I wanted to show that the behaviour is still incorrect. Is there a way to mark a unittest as failing because of outstanding issues? I would like to add support for (r)find in the future. But it requires more research because of the deep call graph of these functions.

eryksun replied:
> I would like to add support for (r)find in the future. But it requires more research because of the deep call graph of these functions.

The exception includes the faulting address. Doesn't it suffice to check that the in-page error or access violation occurred in the mapped range of the mmap instance?

zooba replied:
No, because we don't know what state has been set up that is being bypassed by the exception handler, so we can't safely continue (but maybe we will know once the call graph has been untangled - I haven't looked at it myself).

That said, we almost certainly need the handler to be written in the mmap module, whether it can be untangled or not. So given a lack of choices, it may not be too hard to decide.

eryksun replied:
> No, because we don't know what state has been set up that is being bypassed by the exception handler, so we can't safely continue

Okay, you mean something like if stringlib allocates temporary buffers on the heap that could lead to a memory leak. That would have to be handled locally with a try-finally termination handler. Once an exception handler is found, the subsequent termination handlers get called before the exception handler is called.

zooba replied:
Right, and any locks that have been taken or released need to be restored, refcounts, etc. We have way too much manually managed state to rely on normal unwinding, unfortunately. If we can target the specific operation (like with memcpy) then it's okay.

eryksun replied:
I don't see an issue like that with `_PyBytes_Find()`, `stringlib_find()`, and `fastsearch()`. It's implemented with automatic variables, no locks, or object references. I think it's okay to implement a safe version of `mmap_gfind()` on Windows. For example:

```c
#if defined(MS_WINDOWS) && !defined(DONT_USE_SEH)
static DWORD
filter_page_exception(EXCEPTION_POINTERS *ptrs, EXCEPTION_RECORD *record)
{
    *record = *ptrs->ExceptionRecord;
    if (record->ExceptionCode == EXCEPTION_IN_PAGE_ERROR ||
        record->ExceptionCode == EXCEPTION_ACCESS_VIOLATION)
    {
        return EXCEPTION_EXECUTE_HANDLER;
    }
    return EXCEPTION_CONTINUE_SEARCH;
}


static DWORD
filter_page_exception_method(mmap_object *self,
                             EXCEPTION_POINTERS *ptrs,
                             EXCEPTION_RECORD *record)
{
    *record = *ptrs->ExceptionRecord;
    if (record->ExceptionCode == EXCEPTION_IN_PAGE_ERROR ||
        record->ExceptionCode == EXCEPTION_ACCESS_VIOLATION)
    {
        uintptr_t address = (uintptr_t)record->ExceptionInformation[1];
        if (address >= (uintptr_t)self->data &&
            address < (uintptr_t)self->data + (uintptr_t)self->size)
        {
            return EXCEPTION_EXECUTE_HANDLER;
        }
    }
    return EXCEPTION_CONTINUE_SEARCH;
}
#endif // defined(MS_WINDOWS) && !defined(DONT_USE_SEH)
```

```c
#if defined(MS_WINDOWS) && !defined(DONT_USE_SEH)
static PyObject *
mmap_gfind(mmap_object *self,
           PyObject *args,
           int reverse)
{
    Py_ssize_t start = self->pos;
    Py_ssize_t end = self->size;
    Py_buffer view;
    Py_ssize_t index;

    CHECK_VALID(NULL);
    if (!PyArg_ParseTuple(args, reverse ? "y*|nn:rfind" : "y*|nn:find",
                          &view, &start, &end))
    {
        return NULL;
    }

    if (start < 0) {
        start += self->size;
    }
    if (start < 0) {
        start = 0;
    }
    else if (start > self->size) {
        start = self->size;
    }

    if (end < 0) {
        end += self->size;
    }
    if (end < 0) {
        end = 0;
    }
    else if (end > self->size) {
        end = self->size;
    }

    CHECK_VALID_OR_RELEASE(NULL, view);

    DWORD error = 0;
    EXCEPTION_RECORD record;

    __try {
        if (end < start) {
            index = -1;
        }
        else if (reverse) {
            assert(0 <= start && start <= end && end <= self->size);
            index = _PyBytes_ReverseFind(self->data + start, end - start,
                                         view.buf, view.len, start);
        }
        else {
            assert(0 <= start && start <= end && end <= self->size);
            index = _PyBytes_Find(self->data + start, end - start,
                                  view.buf, view.len, start);
        }
    }
    __except (filter_page_exception_method(
                self, GetExceptionInformation(), &record))
    {
        assert(record.ExceptionCode == EXCEPTION_IN_PAGE_ERROR ||
               record.ExceptionCode == EXCEPTION_ACCESS_VIOLATION);
        if (record.ExceptionCode == EXCEPTION_IN_PAGE_ERROR) {
            error = LsaNtStatusToWinError(
                        (NTSTATUS)record.ExceptionInformation[2]);
        }
        else {
            error = ERROR_NOACCESS;
        }
    }

    PyBuffer_Release(&view);
    if (error) {
        return PyErr_SetFromWindowsErr(error);
    }
    return PyLong_FromSsize_t(index);
}

#else // !defined(MS_WINDOWS) || defined(DONT_USE_SEH)
static PyObject *
mmap_gfind(mmap_object *self,
           PyObject *args,
           int reverse)
{
    Py_ssize_t start = self->pos;
    Py_ssize_t end = self->size;
    Py_buffer view;
    Py_ssize_t index;

    CHECK_VALID(NULL);
    if (!PyArg_ParseTuple(args, reverse ? "y*|nn:rfind" : "y*|nn:find",
                          &view, &start, &end))
    {
        return NULL;
    }

    if (start < 0) {
        start += self->size;
    }
    if (start < 0) {
        start = 0;
    }
    else if (start > self->size) {
        start = self->size;
    }

    if (end < 0) {
        end += self->size;
    }
    if (end < 0) {
        end = 0;
    }
    else if (end > self->size) {
        end = self->size;
    }

    CHECK_VALID_OR_RELEASE(NULL, view);

    if (end < start) {
        index = -1;
    }
    else if (reverse) {
        assert(0 <= start && start <= end && end <= self->size);
        index = _PyBytes_ReverseFind(self->data + start, end - start,
                                     view.buf, view.len, start);
    }
    else {
        assert(0 <= start && start <= end && end <= self->size);
        index = _PyBytes_Find(self->data + start, end - start,
                              view.buf, view.len, start);
    }

    PyBuffer_Release(&view);
    return PyLong_FromSsize_t(index);
}
#endif // defined(MS_WINDOWS) && !defined(DONT_USE_SEH)
```

zooba commented on Lib/test/test_mmap.py:
```diff
@@ -1058,6 +1060,91 @@ def __exit__(self, exc_type, exc_value, traceback):
+
+                m = mmap.mmap(f.fileno(), PAGESIZE)
+                m._protect(PAGE_NOACCESS, 0, PAGESIZE)
+                with suppress(OSError):
```
Pretty sure `suppress` will silently continue if the exception is not raised. Do we want to require that it be raised?

Adding an `assert False, "m.read(...) did not raise"` after the function that should raise ought to handle it.

If we want to pass silently even when the failure is not triggered, then add a comment here mentioning that no exception or OSError are equally acceptable.

Dobatymo replied:
it should be triggered yes, previously I used assertRaises, but I simplified it for the subprocess method
I will add some asserts

Dobatymo committed:
add asserts to tests

zooba commented on Lib/test/test_mmap.py:
```diff
@@ -1058,6 +1060,105 @@ def __exit__(self, exc_type, exc_value, traceback):
+            PAGESIZE = mmap.PAGESIZE
+            PAGE_NOACCESS = 0x01
+
+            with open(TESTFN, 'bw+') as f:
```
This `TESTFN` is remaining after the process finishes (so is the other one), and we require tests clean up after themselves.

I'd suggest changing this one to `sys.argv[1]` (you'll need to import sys), and then pass `TESTFN` in the `assert_python_ok` call. Wrap that call in a try/finally and `os.unlink(TESTFN)` in its own `try`/`except OSError` and we should be good (yes, this is a lot of indenting... I'm not sure we have anything neater...)

Dobatymo replied:
according to eryksun we can just use the `setUp` and `tearDown` handlers right?

EDIT: Because `TESTFN` depends on the process id, I pass it to the subprocess, but the normal `setUp` and `tearDown` handlers are taking care of it afterwards.

eryksun commented on Lib/test/test_mmap.py:
```diff
@@ -1058,6 +1060,105 @@ def __exit__(self, exc_type, exc_value, traceback):
+            from contextlib import suppress
+
+            if os.path.exists(TESTFN):
+                os.unlink(TESTFN)
```
```suggestion
```

`MmapTests.setUp()` is called before each test. It deletes `TESTFN` if it exists. The test file also gets deleted by `MmapTests.tearDown()` after each test, whether the test passes or fails.

Dobatymo committed:
Update Lib/test/test_mmap.py

Co-authored-by: Eryk Sun <eryksun@gmail.com>

Dobatymo committed:
support mmap_gfind

fix test cleanup

Dobatymo committed:
aesthetics

eryksun commented on Modules/mmapmodule.c:
```diff
@@ -255,6 +256,192 @@ do {                                                                    \
 } while (0)
 #endif /* UNIX */
 
+#if defined(MS_WIN32) && !defined(DONT_USE_SEH)
```
```suggestion
#if defined(MS_WINDOWS) && !defined(DONT_USE_SEH)
```

eryksun commented on Modules/mmapmodule.c:
```diff
@@ -255,6 +256,192 @@ do {                                                                    \
+}
+#endif
+
+#if defined(MS_WIN32) && !defined(DONT_USE_SEH)
```
```suggestion
#if defined(MS_WINDOWS) && !defined(DONT_USE_SEH)
```

eryksun commented on Modules/mmapmodule.c:
```diff
@@ -255,6 +256,192 @@ do {                                                                    \
+} while (0)
+#endif
+
+#if defined(MS_WIN32) && !defined(DONT_USE_SEH)
```
```suggestion
#if defined(MS_WINDOWS) && !defined(DONT_USE_SEH)
```

eryksun commented on Modules/mmapmodule.c:
```diff
@@ -255,6 +256,192 @@ do {                                                                    \
+}
+
+static DWORD
+filter_page_exception_method(mmap_object *self, EXCEPTION_POINTERS *ptrs, EXCEPTION_RECORD *record)
```
```suggestion
filter_page_exception_method(mmap_object *self, EXCEPTION_POINTERS *ptrs,
                             EXCEPTION_RECORD *record)
```

PEP 7:
* No line should be longer than 79 characters.
* Breaking long lines: if you can, break after commas in the outermost argument list. Always indent continuation lines appropriately

eryksun commented on Modules/mmapmodule.c:
```diff
@@ -255,6 +256,192 @@ do {                                                                    \
+{
+    *record = *ptrs->ExceptionRecord;
+    if (record->ExceptionCode == EXCEPTION_IN_PAGE_ERROR
+        || record->ExceptionCode == EXCEPTION_ACCESS_VIOLATION) {
```
```suggestion
    if (record->ExceptionCode == EXCEPTION_IN_PAGE_ERROR ||
        record->ExceptionCode == EXCEPTION_ACCESS_VIOLATION)
    {
```

PEP 7:
* When you break a long expression at a binary operator, the operator goes at the end of the previous line

eryksun commented on Modules/mmapmodule.c:
```diff
@@ -255,6 +256,192 @@ do {                                                                    \
+
+        ULONG_PTR address = record->ExceptionInformation[1];
+        if (address >= (ULONG_PTR) self->data
+            && address < (ULONG_PTR) self->data + (ULONG_PTR) self->size)
```
```suggestion
        if (address >= (ULONG_PTR) self->data &&
            address < (ULONG_PTR) self->data + (ULONG_PTR) self->size)
```

eryksun commented on Modules/mmapmodule.c:
```diff
@@ -255,6 +256,192 @@ do {                                                                    \
+    }                                                                      \
+    __except (filter_page_exception(GetExceptionInformation(), &record)) { \
+        assert(record.ExceptionCode == EXCEPTION_IN_PAGE_ERROR             \
+               || record.ExceptionCode == EXCEPTION_ACCESS_VIOLATION);     \
```
```suggestion
        assert(record.ExceptionCode == EXCEPTION_IN_PAGE_ERROR ||          \
               record.ExceptionCode == EXCEPTION_ACCESS_VIOLATION);        \
```

eryksun commented on Modules/mmapmodule.c:
```diff
@@ -255,6 +256,192 @@ do {                                                                    \
+    }                                                                                   \
+    __except (filter_page_exception_method(self, GetExceptionInformation(), &record)) { \
+        assert(record.ExceptionCode == EXCEPTION_IN_PAGE_ERROR                          \
+               || record.ExceptionCode == EXCEPTION_ACCESS_VIOLATION);                  \
```
```suggestion
        assert(record.ExceptionCode == EXCEPTION_IN_PAGE_ERROR ||                       \
               record.ExceptionCode == EXCEPTION_ACCESS_VIOLATION);                     \
```

eryksun commented on Modules/mmapmodule.c:
```diff
@@ -255,6 +256,192 @@ do {                                                                    \
+}
+
+int
+safe_copy_from_slice(char *dest, const char *src, Py_ssize_t start, Py_ssize_t step, Py_ssize_t slicelen) {
```
```suggestion
safe_copy_from_slice(char *dest, const char *src, Py_ssize_t start,
                     Py_ssize_t step, Py_ssize_t slicelen)
{
```

eryksun commented on Modules/mmapmodule.c:
```diff
@@ -255,6 +256,192 @@ do {                                                                    \
+}
+
+int
+safe_copy_to_slice(char *dest, const char *src, Py_ssize_t start, Py_ssize_t step, Py_ssize_t slicelen) {
```
```suggestion
safe_copy_to_slice(char *dest, const char *src, Py_ssize_t start,
                   Py_ssize_t step, Py_ssize_t slicelen)
{
```

eryksun commented on Modules/mmapmodule.c:
```diff
@@ -255,6 +256,192 @@ do {                                                                    \
+#endif
+
+int
+safe_memcpy(void *restrict dest, const void *restrict src, size_t count) {
```
```suggestion
safe_memcpy(void *restrict dest, const void *restrict src, size_t count)
{
```

eryksun commented on Modules/mmapmodule.c:
```diff
@@ -255,6 +256,192 @@ do {                                                                    \
+}
+
+int
+safe_byte_copy(char *dest, const char *src) {
```
```suggestion
safe_byte_copy(char *dest, const char *src)
{
```

eryksun commented on Modules/mmapmodule.c:
```diff
@@ -255,6 +256,192 @@ do {                                                                    \
+}
+
+int
+safe_memchr(char **out, const void *ptr, int ch, size_t count) {
```
```suggestion
safe_memchr(char **out, const void *ptr, int ch, size_t count)
{
```

eryksun commented on Modules/mmapmodule.c:
```diff
@@ -255,6 +256,192 @@ do {                                                                    \
+}
+
+int
+safe_memmove(void *dest, const void *src, size_t count) {
```
```suggestion
safe_memmove(void *dest, const void *src, size_t count)
{
```

eryksun commented on Modules/mmapmodule.c:
```diff
@@ -255,6 +256,192 @@ do {                                                                    \
+int
+_safe_PyBytes_Find(Py_ssize_t *out, mmap_object *self, const char *haystack,
+                   Py_ssize_t len_haystack, const char *needle, Py_ssize_t len_needle,
+                   Py_ssize_t offset) {
```
```suggestion
_safe_PyBytes_Find(Py_ssize_t *out, mmap_object *self, const char *haystack,
                   Py_ssize_t len_haystack, const char *needle,
                   Py_ssize_t len_needle, Py_ssize_t offset)
{
```

eryksun commented on Modules/mmapmodule.c:
```diff
@@ -255,6 +256,192 @@ do {                                                                    \
+int
+_safe_PyBytes_ReverseFind(Py_ssize_t *out, mmap_object *self, const char *haystack,
+                          Py_ssize_t len_haystack, const char *needle, Py_ssize_t len_needle,
+                          Py_ssize_t offset) {
```
```suggestion
_safe_PyBytes_ReverseFind(Py_ssize_t *out, mmap_object *self,
                          const char *haystack, Py_ssize_t len_haystack,
                          const char *needle, Py_ssize_t len_needle,
                          Py_ssize_t offset)
{
```

eryksun commented on Modules/mmapmodule.c:
```diff
@@ -255,6 +256,192 @@ do {                                                                    \
+                          Py_ssize_t len_haystack, const char *needle, Py_ssize_t len_needle,
+                          Py_ssize_t offset) {
+    HANDLE_INVALID_MEM_METHOD(self,
+        *out = _PyBytes_ReverseFind(haystack, len_haystack, needle, len_needle, offset);
```
```suggestion
        *out = _PyBytes_ReverseFind(haystack, len_haystack, needle, len_needle,
                                    offset);
```

eryksun commented on Modules/mmapmodule.c:
```diff
@@ -1123,24 +1381,21 @@ mmap_ass_subscript(mmap_object *self, PyObject *item, PyObject *value)
-                 cur += step, i++)
-            {
-                self->data[cur] = ((char *)vbuf.buf)[i];
+            if (safe_copy_from_slice(self->data, (char *)vbuf.buf, start, step, slicelen) < 0) {
```
```suggestion
            if (safe_copy_from_slice(self->data, (char *)vbuf.buf, start, step,
                                     slicelen) < 0)
            {
```

eryksun commented on Modules/mmapmodule.c:
```diff
@@ -255,6 +256,192 @@ do {                                                                    \
+{
+    *record = *ptrs->ExceptionRecord;
+    if (record->ExceptionCode == EXCEPTION_IN_PAGE_ERROR
+        || record->ExceptionCode == EXCEPTION_ACCESS_VIOLATION) {
```
```suggestion
    if (record->ExceptionCode == EXCEPTION_IN_PAGE_ERROR ||
        record->ExceptionCode == EXCEPTION_ACCESS_VIOLATION)
    {
```

eryksun commented on Modules/mmapmodule.c:
```diff
@@ -307,8 +504,11 @@ mmap_read_method(mmap_object *self,
-    result = PyBytes_FromStringAndSize(&self->data[self->pos], num_bytes);
-    self->pos += num_bytes;
+
+    PyObject *result = _safe_PyBytes_FromStringAndSize(self->data + self->pos, num_bytes);
```
```suggestion
    PyObject *result = _safe_PyBytes_FromStringAndSize(self->data + self->pos,
                                                       num_bytes);
```

eryksun commented on Modules/mmapmodule.c:
```diff
@@ -341,25 +541,36 @@ mmap_gfind(mmap_object *self,
+            if (_safe_PyBytes_ReverseFind(&index, self,
                 self->data + start, end - start,
-                view.buf, view.len, start);
+                view.buf, view.len, start) < 0) {
```
```suggestion
                view.buf, view.len, start) < 0)
            {
```

eryksun commented on Modules/mmapmodule.c:
```diff
@@ -341,25 +541,36 @@ mmap_gfind(mmap_object *self,
+            if (_safe_PyBytes_Find(&index, self,
                 self->data + start, end - start,
-                view.buf, view.len, start);
+                view.buf, view.len, start) < 0) {
```
```suggestion
                view.buf, view.len, start) < 0)
            {
```

eryksun commented on Modules/mmapmodule.c:
```diff
@@ -855,6 +1075,27 @@ mmap__sizeof__method(mmap_object *self, void *Py_UNUSED(ignored))
+        return NULL;
+    }
+
+    if (VirtualProtect((void *) (self->data + start), length, flNewProtect, &flOldProtect) == 0) {
```
```suggestion
    if (!VirtualProtect((void *) (self->data + start), length, flNewProtect,
                        &flOldProtect))
    {
```

eryksun commented on Modules/mmapmodule.c:
```diff
@@ -1001,23 +1255,20 @@ mmap_subscript(mmap_object *self, PyObject *item)
-            for (cur = start, i = 0; i < slicelen;
-                 cur += step, i++) {
-                result_buf[i] = self->data[cur];
+            if (safe_copy_to_slice(result_buf, self->data, start, step, slicelen) < 0) {
```
```suggestion
            if (safe_copy_to_slice(result_buf, self->data, start, step,
                                   slicelen) < 0)
            {
```

Dobatymo committed:
Apply suggestions from code review

Co-authored-by: Eryk Sun <eryksun@gmail.com>

eryksun commented on Modules/mmapmodule.c:
```diff
@@ -255,6 +256,207 @@ do {                                                                    \
+#else
+#define HANDLE_INVALID_MEM_METHOD(self, sourcecode)                                     \
+do {                                                                                    \
+    sourcecode                                                                          \
```
```suggestion
#define HANDLE_INVALID_MEM_METHOD(self, sourcecode)                           \
do {                                                                          \
    EXCEPTION_RECORD record;                                                  \
    __try {                                                                   \
        sourcecode                                                            \
    }                                                                         \
    __except (filter_page_exception_method(self, GetExceptionInformation(),   \
                                           &record)) {                        \
        assert(record.ExceptionCode == EXCEPTION_IN_PAGE_ERROR ||             \
               record.ExceptionCode == EXCEPTION_ACCESS_VIOLATION);           \
        if (record.ExceptionCode == EXCEPTION_IN_PAGE_ERROR) {                \
            NTSTATUS status = (NTSTATUS) record.ExceptionInformation[2];      \
            ULONG code = LsaNtStatusToWinError(status);                       \
            PyErr_SetFromWindowsErr(code);                                    \
        }                                                                     \
        else if (record.ExceptionCode == EXCEPTION_ACCESS_VIOLATION) {        \
            PyErr_SetFromWindowsErr(ERROR_NOACCESS);                          \
        }                                                                     \
        return -1;                                                            \
    }                                                                         \
} while (0)
#else
#define HANDLE_INVALID_MEM_METHOD(self, sourcecode)                           \
do {                                                                          \
    sourcecode                                                                \
```

I missed this earlier. It's just a superficial change to comply with the PEP 7 line-length limit of 79 characters.

Dobatymo committed:
remove restrict for compliance with the python C dialect

Dobatymo committed:
Update Modules/mmapmodule.c

Co-authored-by: Eryk Sun <eryksun@gmail.com>
